### PR TITLE
retry converge on exit status

### PR DESCRIPTION
**Pull request by jeremymv2**
This allows folks to better test an upgrade in Test Kitchen when `kill_chef_on_upgrade` is true (default).   It allows Test Kitchen to capture the exit code from the `'omnibus chef killer'` ruby_block without bubbling up the error, then run another converge, this time with the newly installed chef-client!

You would add this to your .kitchen.yml
```yaml
provisioner:
  name: chef_zero
  retry_on_exit_code:
    - 3
  max_retries: 1
  wait_for_retry: 1
  client_rb:
    exit_status: :enabled
    client_fork: false
```

Then kitchen retries the converge:
```
~/Devel/ChefProject/omnibus_updater_wrapper (master *%)$ cat recipes/default.rb
#
# Cookbook Name:: omnibus_updater_wrapper
# Recipe:: default
#
# Copyright (c) 2016 The Authors, All Rights Reserved.

include_recipe 'omnibus_updater::default'

```
```
~/Devel/ChefProject/omnibus_updater_wrapper (master *%)$ kitchen destroy; kitchen converge
-----> Starting Kitchen (v1.11.1)
-----> Destroying <default-ubuntu-1404>...
       ==> default: Forcing shutdown of VM...
       ==> default: Destroying VM and associated drives...
       Vagrant instance <default-ubuntu-1404> destroyed.
       Finished destroying <default-ubuntu-1404> (0m3.59s).
-----> Kitchen is finished. (0m5.22s)
-----> Starting Kitchen (v1.11.1)
-----> Creating <default-ubuntu-1404>...
       Bringing machine 'default' up with 'virtualbox' provider...
       ==> default: Importing base box 'bento/ubuntu-14.04'...
==> default: Matching MAC address for NAT networking...
       ==> default: Checking if box 'bento/ubuntu-14.04' is up to date...
       ==> default: A newer version of the box 'bento/ubuntu-14.04' is available! You currently
       ==> default: have version '2.2.7'. The latest is version '2.2.9'. Run
       ==> default: `vagrant box update` to update.
       ==> default: Setting the name of the VM: kitchen-omnibus_updater_wrapper-default-ubuntu-1404_default_1474412782467_90416
       ==> default: Clearing any previously set network interfaces...
       ==> default: Preparing network interfaces based on configuration...
           default: Adapter 1: nat
       ==> default: Forwarding ports...
           default: 22 (guest) => 2222 (host) (adapter 1)
       ==> default: Booting VM...
       ==> default: Waiting for machine to boot. This may take a few minutes...
           default: SSH address: 127.0.0.1:2222
           default: SSH username: vagrant
           default: SSH auth method: private key
           default:
           default: Vagrant insecure key detected. Vagrant will automatically replace
           default: this with a newly generated keypair for better security.
           default:
           default: Inserting generated public key within guest...
           default: Removing insecure key from the guest if it's present...
           default: Key inserted! Disconnecting and reconnecting using new SSH key...
       ==> default: Machine booted and ready!
       ==> default: Checking for guest additions in VM...
       ==> default: Setting hostname...
       ==> default: Machine not provisioned because `--no-provision` is specified.
       [SSH] Established
       Vagrant instance <default-ubuntu-1404> created.
       Finished creating <default-ubuntu-1404> (0m36.06s).
-----> Converging <default-ubuntu-1404>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 4.3.5...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
-----> Installing Chef Omnibus (12.11)
       Downloading https://omnitruck.chef.io/install.sh to file /tmp/install.sh
       Trying wget...
       Download complete.
       ubuntu 14.04 x86_64
       Getting information for chef stable 12.11 for ubuntu...
       downloading https://omnitruck.chef.io/stable/chef/metadata?v=12.11&p=ubuntu&pv=14.04&m=x86_64
         to file /tmp/install.sh.1404/metadata.txt
       trying wget...
       sha1	bc4a6642d6093b54de856da31f0651a56b0fdc8e
       sha256	f1cf5d0f6dd12d2d2296ec6d8dbb16363f8541f5c15298cafa70e65ff2b5a22f
       url	https://packages.chef.io/stable/ubuntu/14.04/chef_12.11.18-1_amd64.deb
       version	12.11.18
       downloaded metadata file looks valid...
       downloading https://packages.chef.io/stable/ubuntu/14.04/chef_12.11.18-1_amd64.deb
         to file /tmp/install.sh.1404/chef_12.11.18-1_amd64.deb
       trying wget...
       Comparing checksum with sha256sum...
       Installing chef 12.11
       installing with dpkg...
       Selecting previously unselected package chef.
(Reading database ... 32948 files and directories currently installed.)
       Preparing to unpack .../chef_12.11.18-1_amd64.deb ...
       Unpacking chef (12.11.18-1) ...
       Setting up chef (12.11.18-1) ...
       Thank you for installing Chef!
       Transferring files to <default-ubuntu-1404>
       Starting Chef Client, version 12.11.18
       Creating a new client identity for default-ubuntu-1404 using the validator key.
       resolving cookbooks for run list: ["omnibus_updater_wrapper::default"]
       Synchronizing Cookbooks:
         - omnibus_updater (2.0.0)
         - omnibus_updater_wrapper (0.1.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 7 resources
       Recipe: omnibus_updater::downloader
         * directory[/tmp/kitchen/cache/omnibus_updater] action create
           - create new directory /tmp/kitchen/cache/omnibus_updater
         * remote_file[omnibus_remote[chef_12.14.77-1_amd64.deb]] action create_if_missing
           - create new file /tmp/kitchen/cache/omnibus_updater/chef_12.14.77-1_amd64.deb
           - update content in file /tmp/kitchen/cache/omnibus_updater/chef_12.14.77-1_amd64.deb from none to 281c80
           (file sizes exceed 10000000 bytes, diff output suppressed)
       Recipe: omnibus_updater::installer
         * ruby_block[Omnibus Chef install notifier] action create
           - execute the ruby block Omnibus Chef install notifier
         * service[chef-client-omnibus] action nothing (skipped due to action :nothing)
         * file[/tmp/nocheck] action create (skipped due to only_if)
         * ruby_block[omnibus chef killer] action nothing (skipped due to action :nothing)
         * execute[omnibus_install[chef_12.14.77-1_amd64.deb]] action nothing (skipped due to action :nothing)
         * ruby_block[Omnibus Chef install notifier] action nothing (skipped due to action :nothing)
         * execute[omnibus_install[chef_12.14.77-1_amd64.deb]] action run
           - execute dpkg -i /tmp/kitchen/cache/omnibus_updater/chef_12.14.77-1_amd64.deb
         * ruby_block[omnibus chef killer] action create[2016-09-20T23:07:22+00:00] FATAL: New omnibus chef version installed. Killing Chef run!


           ================================================================================
           Error executing action `create` on resource 'ruby_block[omnibus chef killer]'
           ================================================================================

           SystemExit
           ----------
           exit

           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/omnibus_updater/recipes/installer.rb:67:in `block (2 levels) in from_file'

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/omnibus_updater/recipes/installer.rb

            65:   ruby_block 'omnibus chef killer' do
            66:     block do
            67:       Chef::Application.fatal!('New omnibus chef version installed. Killing Chef run!', node['omnibus_updater']['kill_chef_on_upgrade_exit_code'])
            68:     end
            69:     action :nothing
            70:     only_if { node['omnibus_updater']['kill_chef_on_upgrade'] }
            71:   end
            72:

           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/omnibus_updater/recipes/installer.rb:65:in `from_file'

           ruby_block("omnibus chef killer") do
             action [:nothing]
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             block_name "omnibus chef killer"
             declared_type :ruby_block
             cookbook_name "omnibus_updater"
             recipe_name "installer"
             block #<Proc:0x00000004be3cb0@/tmp/kitchen/cache/cookbooks/omnibus_updater/recipes/installer.rb:66>
             only_if { #code block }
           end

           Platform:
           ---------
           x86_64-linux


       Running handlers:
       [2016-09-20T23:07:22+00:00] ERROR: Running exception handlers
       Running handlers complete
       [2016-09-20T23:07:22+00:00] ERROR: Exception handlers complete
       Chef Client failed. 4 resources updated in 16 seconds
       [2016-09-20T23:07:22+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2016-09-20T23:07:22+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report

       Starting Chef Client, version 12.14.77
       resolving cookbooks for run list: ["omnibus_updater_wrapper::default"]
       Synchronizing Cookbooks:
         - omnibus_updater_wrapper (0.1.0)
         - omnibus_updater (2.0.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 7 resources
       Recipe: omnibus_updater::downloader
         * directory[/tmp/kitchen/cache/omnibus_updater] action create (up to date)
         * remote_file[omnibus_remote[chef_12.14.77-1_amd64.deb]] action create_if_missing (skipped due to only_if)
       Recipe: omnibus_updater::installer
         * service[chef-client-omnibus] action nothing (skipped due to action :nothing)
         * file[/tmp/nocheck] action create (skipped due to only_if)
         * ruby_block[omnibus chef killer] action nothing (skipped due to action :nothing)
         * execute[omnibus_install[chef_12.14.77-1_amd64.deb]] action nothing (skipped due to action :nothing)
         * ruby_block[Omnibus Chef install notifier] action nothing (skipped due to action :nothing)

       Running handlers:
       Running handlers complete
       Chef Client finished, 0/7 resources updated in 01 seconds
       Finished converging <default-ubuntu-1404> (0m36.08s).
-----> Kitchen is finished. (1m13.73s)
```

### CODE REVIEW

**Comment by tas50** [comment]
I like this idea and to be honest I didn't realize we could have kitchen retry, which is a pretty handy trick. I do think we should change this up a bit though. It would be nice to exit using an official exit code for upgrades. I opened a PR to modify the exit status RFC to add code 213 for upgrades. If / when that gets accepted it would be nice to default to code 213 and thus eliminate the need for some of the warning around the exit codes in the RFC.

**Comment by jeremymv2** [comment]
@tas50 I will, but first we need to add the new exit code to chef-client
https://github.com/chef/chef/pull/5428

**Comment by jeremymv2** [comment]
Now updated with exit code 213

File: README.md [hunk 1]
```diff
@@ -59,6 +59,22 @@ Force installation of the latest version regardless of value stored in version a
...
+    - 213
+  max_retries: 1
+  wait_for_retry: 1
+  client_rb:
+    exit_status: :enabled
+    client_fork: false
+```
+
+exit code 213 is a special designation for Chef Client Upgrade
+https://github.com/chef/chef-rfc/blob/master/rfc062-exit-status.md
```
Comment (jeremymv2) [inline README.md]: @lamont-granquist thanks!
I added a blurb about it requiring >= 12.15.19

**Comment by jeremymv2** [comment]
@tas50 @lamont-granquist I've tested this with the release candidate `12.15.19` and it worked as expected.

**Pull request merged**